### DISCUSSION TITLE
OpenDDSWrapper (or depenency) not found.

### DISCUSSION THREAD

**Discussion by jrobideaux** [jmmorato/openddsharp#134] [Q&A] [answered]
Hello everyone,

I've been following along w/ the instructions posted here: [Getting Started](https://www.openddsharp.com/articles/getting_started.html)
I've managed to get the IDL and Publisher projects to build with no errors.
However, when I run the Publisher app, I get the following error message.
Looks to be happening right at the call to Ace.Init().

`C:\Projects\Publisher\bin\Debug\net6.0\win-x64>Publisher.exe
Unhandled exception. System.DllNotFoundException: Unable to load DLL 'OpenDDSWrapper' or one of its dependencies: The specified
module could not be found. (0x8007007E)
   at OpenDDSharp.Ace.UnsafeNativeMethods.NativeInit()
   at OpenDDSharp.Ace.Init()
   at Publisher.Program.Main(String[] args) in C:\Projects\Publisher\Program.cs:line 13`

I have tried with the latest nuget packages (0.8.22297.100), as well as the one cited in the above article (0.8.22127.96)

Any insights or questions would be most appreciated!
Many thanks in advance.
-John

**Comment by jmmorato** [comment]
Have you installed the [C++ Redistributable](https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170) in your system?

**Reply by jrobideaux** [reply]
I do have the "C++ 2022 Redistributable Update" installed.  But I see there's an "MSMs" as well.
I can try adding that one and see what happens.

Guess I should mention that I am using VS 2022.  Do I need to try this with VS 2019, perhaps?

![image](https://user-images.githubusercontent.com/11408814/199578508-00a6cc6c-3995-4a3a-a383-3b9e030ae656.png)

**Reply by jmmorato** [reply]
VS 2022 should be ok

**Comment by jmmorato** [comment]
Try to build using directly the dotnet CLI:
`dotnet clean Publisher.csproj`
`dotnet build Publisher.csproj --runtime win-x64 --no-self-contained`

**Reply by jrobideaux** [reply]
Yup..  I've done the following from the command line, as well as opening up the VS Command Prompt window and trying it there.
----------------------------------------------

C:\Projects>cd Publisher
C:\Projects\Publisher>dotnet clean Publisher.csproj

Microsoft (R) Build Engine version 17.2.0+41abc5629 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.
Build started 11/2/2022 2:16:18 PM.
Build succeeded.
    0 Warning(s)
    0 Error(s)
Time Elapsed 00:00:00.81

C:\Projects\Publisher>dotnet build Publisher.csproj --runtime win-x64 --no-self-contained

Microsoft (R) Build Engine version 17.2.0+41abc5629 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.
  Determining projects to restore...
  All projects are up-to-date for restore.
  Publisher -> C:\Projects\Publisher\bin\Debug\net6.0\win-x64\Publisher.dll
Build succeeded.
    0 Warning(s)
    0 Error(s)
Time Elapsed 00:00:01.32

C:\Projects\Publisher>cd bin\Debug\net6.0\win-x64

C:\Projects\Publisher\bin\Debug\net6.0\win-x64>Publisher.exe
Unhandled exception. System.DllNotFoundException: Unable to load DLL "OpenDDSWrapper" or one of its dependencies: The specified
module could not be found. (0x8007007E)
   at OpenDDSharp.Ace.UnsafeNativeMethods.NativeInit()
   at OpenDDSharp.Ace.Init()
   at Publisher.Program.Main(String[] args) in C:\Projects\Publisher\Program.cs:line 13

C:\Projects\Publisher\bin\Debug\net6.0\win-x64>dotnet Publisher.dll
Unhandled exception. System.DllNotFoundException: Unable to load DLL "OpenDDSWrapper" or one of its dependencies: The specified
module could not be found. (0x8007007E)
   at OpenDDSharp.Ace.UnsafeNativeMethods.NativeInit()
   at OpenDDSharp.Ace.Init()
   at Publisher.Program.Main(String[] args) in C:\Projects\Publisher\Program.cs:line 13

**Reply by jmmorato** [reply]
Please execute `dumpbin /dependents OpenDDSharpWrapper.dll` and paste the output here, perhaps we find some clues

**Reply by jrobideaux** [reply]
C:\Projects\Publisher\bin\Debug\net6.0\win-x64>dumpbin /dependents OpenDDSWrapper.dll
Microsoft (R) COFF/PE Dumper Version 14.32.31332.0
Copyright (C) Microsoft Corporation.  All rights reserved.

Dump of file OpenDDSWrapper.dll
File Type: DLL

Image has the following dependencies:
    OpenDDS_Multicast.dll
    OpenDDS_Rtps_Udp.dll
    OpenDDS_Rtps.dll
    OpenDDS_InfoRepoDiscovery.dll
    OpenDDS_Shmem.dll
    OpenDDS_Tcp.dll
    OpenDDS_Udp.dll
    OpenDDS_Dcps.dll
    **TAO_Valuetype.dll**
    TAO.dll
    ACE.dll
    MSVCP140.dll
    VCRUNTIME140.dll
    VCRUNTIME140_1.dll
    api-ms-win-crt-runtime-l1-1-0.dll
    api-ms-win-crt-heap-l1-1-0.dll
    KERNEL32.dll

  Summary
        3000 .data
        2000 .pdata
       18000 .rdata
        1000 .reloc
        1000 .rsrc
       23000 .text

**Reply by jrobideaux** [reply]
Looking in the folder, aside from the standard MS assemblies, I see all of the above - **_Except_** - for "TAO_Valuetype.dll".
This is what I have:

```
TAO_PortableServer.dll
TAO_PI.dll
TAO_CodecFactory.dll
TAO_BiDirGIOP.dll
TAO_AnyTypeCode.dll
TAO.dll
ACE.dll
OpenDDS_Tcp.dll
OpenDDS_Shmem.dll
OpenDDS_Rtps_Udp.dll
OpenDDS_Rtps.dll
OpenDDS_Multicast.dll
OpenDDS_InfoRepoDiscovery.dll
OpenDDS_Dcps.dll
OpenDDS_Udp.dll
OpenDDSharp.Standard.dll
OpenDDSWrapper.dll
rtps.ini
TestMessage.pdb
TestMessage.dll
Publisher.pdb
Publisher.dll
Publisher.exe
Publisher.deps.json
Publisher.runtimeconfig.json
```

**Reply by jmmorato** [reply]
I have published a new version [0.8.22309.101-beta](https://www.nuget.org/packages/OpenDDSharp.Standard/0.8.22309.101-beta) that includes TAO_Valuetype.dll file.
Please, check it and let me know if works for you.

**Comment by jmmorato** [comment, marked as answer]
Fixed on version 0.8.23023.106
